ferrocene/tools/self-test/src/linkers.rs

```diff
@@ -1,71 +1,257 @@
     BundledLld,
-    GccUbuntu18 { target: &'static str, mode: GccMode },
-}
-
-#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-pub(crate) enum GccMode {
-    Normal,
-    #[cfg_attr(not(test), allow(unused))]
-    BareMetal,
+    /// The system's native C compiler is required
+    HostCC,
+    /// Some kind of cross compiler, with one of the given target prefixes
+    CrossCC(&'static [&'static str]),
 }
 
+/// We need to find a system C compiler and determine what flags should be
+/// added when calling `rustc`.
+///
+/// These flags will include `-Clinker=/path/to/cc`.
 pub(crate) fn check_and_add_rustflags(
     reporter: &dyn Reporter,
     environment: &Environment,
     sysroot: &Path,
     targets: &mut [Target],
 ) -> Result<(), Error> {
-    // Multiple installed targets might use the same linker, deduplicate them.
-    let mut linkers = BTreeMap::new();
-    for target in targets {
-        linkers.entry(target.linker).or_insert_with(Vec::new).push(target);
-    }
-
-    for (linker, targets) in linkers {
-        let triples = targets.iter().map(|t| t.triple.to_string()).collect::<Vec<_>>();
-        let (bin, flavor, extra_flags): (_, _, &[&str]) = match linker {
-            Linker::BundledLld => (check_bundled_lld(reporter, sysroot, &triples)?, "ld.lld", &[]),
-            Linker::GccUbuntu18 { target: gcc_target, mode } => (
-                check_gcc(reporter, environment, &triples, gcc_target, mode, [7, 5])?,
-                "gcc",
-                match mode {
-                    GccMode::Normal => &[],
-                    GccMode::BareMetal => &["-ffreestanding", "-nostdlib"],
-                },
-            ),
+    // Step 1. Check we have ld.lld available
+    let _rust_lld_path = find_bundled_lld(reporter, sysroot)?;
+    let lld_bin = find_bundled_lld_wrapper(reporter, sysroot)?;
+    let lld_dir = lld_bin.parent().expect("ld.lld should be a in a directory");
+
+    // Step 2. Check the C compiler works on each target that needs one
+    'target_loop: for target in targets {
+        let prefix_list: &[&str] = match target.linker {
+            Linker::BundledLld => {
+                reporter
+                    .skipped(&format!("Target `{}` does not require a C compiler", target.triple));
+                continue 'target_loop;
+            }
+            Linker::HostCC => &[""],
+            Linker::CrossCC(list) => list,
         };
-        let bin = bin.to_str().ok_or_else(|| Error::NonUtf8Path { path: bin.clone() })?;
-        for target in targets {
-            target.rustflags.push(format!("-Clinker={bin}"));
-            target.rustflags.push(format!("-Clinker-flavor={flavor}"));
-            for flag in extra_flags {
-                target.rustflags.push(format!("-Clink-arg={flag}"));
+        for cc_prefix in prefix_list {
+            for compiler_kind in ["cc", "gcc", "clang"] {
+                let temp_dir = tempfile::tempdir().expect("making temp dir");
```

Comment:
There is an error variant for this already, `Error::TemporaryCompilationDirectoryCreationFailed`. We also create a temporary directory in `compile.rs`, it might be worth to extract that to `main()` and pass the path to the temporary directory to all functions requiring one?

Reply:
I wanted each compiler test to have its own temp directory, so we don't have to worry about output files lying around from the previous test.

But I will use the correct error code.